dotnet/src/Connectors/Connectors.Memory.Redis/RedisVectorRecordStore.cs

```diff
@@ -0,0 +1,399 @@
+    {
+        Verify.NotNull(keys);
+        var keysList = keys.ToList();
+
+        // Create Options
+        var collectionName = this.ChooseCollectionName(options?.CollectionName);
+        var maybePrefixedKeys = keysList.Select(key => this.PrefixKeyIfNeeded(key, collectionName));
+        var redisKeys = maybePrefixedKeys.Select(x => new RedisKey(x)).ToArray();
+
+        // Get the list of redis results.
+        var redisResults = await RunOperationAsync(
+            collectionName,
+            "MGET",
+            () => this._database
+                .JSON()
+                .MGetAsync(redisKeys, "$")).ConfigureAwait(false);
+
+        // Loop through each key and result and convert to the caller's data model.
+        for (int i = 0; i < keysList.Count; i++)
+        {
+            var key = keysList[i];
+            var redisResult = redisResults[i];
+
+            // Check if the key was found before trying to parse the result.
+            if (redisResult.IsNull || redisResult is null)
+            {
+                throw new VectorStoreOperationException($"Could not find document with key '{key}'");
+            }
+
+            // Check if the value contained any json text before trying to parse the result.
+            var redisResultString = redisResult.ToString();
+            if (redisResultString is null)
+            {
+                throw new VectorStoreRecordMappingException($"Document with key '{key}' does not contain any json.");
+            }
+
+            // Convert to the caller's data model.
+            yield return RunModelConversion(
+                collectionName,
+                "MGET",
+                () =>
+                {
+                    var node = JsonSerializer.Deserialize<JsonNode>(redisResultString)!;
+                    return this._mapper.MapFromStorageToDataModel((key, node));
+                });
+        }
+    }
+
+    /// <inheritdoc />
+    public Task DeleteAsync(string key, DeleteRecordOptions? options = default, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNullOrWhiteSpace(key);
+
+        // Create Options
+        var collectionName = this.ChooseCollectionName(options?.CollectionName);
+        var maybePrefixedKey = this.PrefixKeyIfNeeded(key, collectionName);
+
+        // Remove.
+        return RunOperationAsync(
+            collectionName,
+            "DEL",
+            () => this._database
+                .JSON()
+                .DelAsync(maybePrefixedKey));
+    }
+
+    /// <inheritdoc />
+    public Task DeleteBatchAsync(IEnumerable<string> keys, DeleteRecordOptions? options = default, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(keys);
+
+        // Remove records in parallel.
+        var tasks = keys.Select(key => this.DeleteAsync(key, options, cancellationToken));
```

Comment:
If deleting one of the records fails then the entire operation fails. Is this what we want?

Reply:
We have a couple of options:
1. We throw regardless of error type on individual records.
2. We throw for errors on individual records, except if error reason is that a record doesn't exist already.
3. We succeed the entire batch.
4. We return per item results.

The problem with 4 is that not all databases support per item results, so that's not an option.  I think 3 is bad, since the user would probably want to retry if the failure is transient.  2 is probably the best, assuming that the database actually supports that behavior if it has native batch support.  IMO it seems more likely and obvious that a delete batch operation would do 2 than 1 anyway.

Thoughts?

Reply:
As discussed offline, I've gone for option 2, which as it happens is already the current behavior, so I've just updated the comments and added integration tests to verify that this works as expected.